scripts/lib/CIME/case/case_run.py

```diff
@@ -106,29 +115,36 @@ def _run_model_impl(case, lid, skip_pnl=False, da_cycle=0):
         stat = run_and_log_case_status(run_func, "model execution", caseroot=case.get_value("CASEROOT"))
         model_logfile = os.path.join(rundir, model + ".log." + lid)
         # Determine if failure was due to a failed node, if so, try to restart
-        if node_fail_re:
-            node_fail_regex = re.compile(node_fail_re)
+        if retry_run_re or node_fail_re:
             model_logfile = os.path.join(rundir, model + ".log." + lid)
             if os.path.exists(model_logfile):
-                num_fails = len(node_fail_regex.findall(open(model_logfile, 'r').read()))
-                if num_fails > 0 and case.spare_nodes >= num_fails:
+                num_node_fails=0
+                num_retry_fails=0
+                if node_fail_re:
+                    num_node_fails = len(node_fail_regex.findall(open(model_logfile, 'r').read()))
+                if retry_run_re:
+                    num_retry_fails = len(retry_run_regex.findall(open(model_logfile, 'r').read()))
+                logger.debug ("RETRY: num_retry_fails {} spare_nodes {} retry_count {}".
+                              format(num_retry_fails, case.spare_nodes, retry_count))
+                if num_node_fails > 0 and case.spare_nodes >= num_node_fails:
                         # We failed due to node failure!
                     logger.warning("Detected model run failed due to node failure, restarting")
-
+                    case.spare_nodes -= num_node_fails
+                    loop = True
+                elif num_retry_fails > 0 and retry_count >= num_retry_fails:
+                    logger.warning("Detected model run failed, restarting")
+                    retry_count -= num_retry_fails
+                    loop = True
+                if loop:
                     # Archive the last consistent set of restart files and restore them
-                    case.case_st_archive(no_resubmit=True)
-                    case.restore_from_archive()
-
+                    if case.get_value("DOUT_S"):
```

Comment:
So, if DOUT_S is off, then how will these retry attempts work at all?

Reply:
If DOUT_S is off the restart files remain in the run directory - in most cases DOUT_S is true.   In cases where DOUT_S is False the archive directory was never created and so you would get an error trying to write there.

Reply:
OK, so I guess this is the best we're going to do here.